harness/determined/pytorch/experimental/_torch_batch_process.py

```diff
@@ -0,0 +1,433 @@
+
+    info = det.get_cluster_info()
+    if info and (len(info.container_addrs) > 1 or len(info.slot_ids) > 1):
+        raise ValueError(
+            "In multi-slot managed cluster training, you must wrap your training script with a "
+            "distributed launch layer such as determined.launch.torch_distributed"
+        )
+    return None
+
+
+def _initialize_default_inference_context() -> core.Context:
+    distributed_context = _initialize_distributed_backend()
+    # Use WorkerAskChief mode to ensure synchronize correctly across worker
+    # Using WorkerAskMaster mode could lead to some workers exiting when others
+    # are waiting for synchronization.
+    return det.core.init(
+        distributed=distributed_context, preempt_mode=core.PreemptMode.WorkersAskChief
+    )
+
+
+class TorchBatchProcessor(metaclass=abc.ABCMeta):
+    """
+    User can initialize necessary resources in the init function, such as
+    - model for prediction
+    - storage client (e.g. s3 client)
+    """
+
+    @abc.abstractmethod
+    def __init__(self, context: TorchBatchProcessorContext) -> None:
+        pass
+
+    @abc.abstractmethod
+    def process_batch(self, batch: Any, batch_idx: int) -> None:
+        pass
+
+    def on_checkpoint_start(self) -> None:  # noqa: B027
+        """
+        Overwrite this function to run certain logic before checkpointing
+        This function will be called right before each checkpoint.
+        """
+        pass
+
+    def on_finish(self) -> None:  # noqa: B027
+        """
+        This function will be called right before exiting after completing iteration
+        over dataset
+        """
+        pass
+
+
+def _load_state(checkpoint_directory: pathlib.Path) -> Any:
+    checkpoint_directory = pathlib.Path(checkpoint_directory)
+    with checkpoint_directory.joinpath("metadata.json").open("r") as f:
+        metadata = json.load(f)
+        return metadata
+
+
+def _synchronize_and_checkpoint(
+    core_context: core.Context, steps_completed: int, rank: int, default_output_uuid: str
+) -> None:
+    """
+    Synchronize the workers and create checkpoint to record steps completed
+    """
+    if rank == 0:
+        steps_completed_list = core_context.distributed.gather(steps_completed)
+        if steps_completed_list is None:
+            return
+        min_steps_completed = min(steps_completed_list)
+
+        checkpoint_metadata = {
+            "steps_completed": min_steps_completed,
+            "default_output_uuid": default_output_uuid,
+        }
+        with core_context.checkpoint.store_path(checkpoint_metadata) as (path, uuid):
+            with open(os.path.join(path, "batch_completed.json"), "w") as file_obj:
+                json.dump({"batch_completed": min_steps_completed}, file_obj)
+    else:
+        core_context.distributed.gather(steps_completed)
+
+
+def _report_progress_to_master(
+    searcher_op: core.DummySearcherOperation,
```

Comment:
why is this typed to a DummySearcherOperation?  Shouldn't it support any kind of SearcherOperation?

Reply:
Discussed offline, we can leave it as is and update after searcher context removal.